Cannot compile generic method on generic struct

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->
I encountered this bug when investigating rustc's mangling scheme.


I tried this code:

```rust
#[lang = "sized"]
pub trait Sized {}

struct S<T>(T);

impl S<i32> {
    fn f<S>(t: S) -> S {
        t
    }
}

pub fn main() {
    S::<i32>::f::<i32>(0);
}

```

I expected to see this happen: 
rustc compiles this code.

Instead, this happened: 

```bash
a.rs:13:19: error: generic item takes at most 1 type arguments but 1 were supplied
    7 |     fn f<S>(t: S) -> S {
      |          ~         
......
   13 |     S::<i32>::f::<i32>(0);
      |                   ^~~
a.rs:13:5: error: Failed to resolve expression of function call
   13 |     S::<i32>::f::<i32>(0);
      |     ^
```

### Meta

- What version of Rust GCC were you using, git sha if possible.

--- a/gcc/testsuite/rust/compile/issue-1773.rs
+++ b/gcc/testsuite/rust/compile/issue-1773.rs
@@ -1,4 +1,5 @@
 #[lang = "sized"]
+// { dg-skip-if "" { *-*-* } }
 pub trait Sized {}
 
 trait Foo<T> {

